File: 11.flyweight/flyweight.java
import java.util.HashMap;
import java.util.Map;

// Flyweight
class Font {
    private final String typeface;

    public Font(String typeface) {
        this.typeface = typeface;
    }

    public String getTypeface() {
        return typeface;
    }
}

// Flyweight Factory
class FontFactory {
    private Map<String, Font> fonts = new HashMap<>();

    public Font getFont(String typeface) {
        if (!fonts.containsKey(typeface)) {
            fonts.put(typeface, new Font(typeface));
        }
        return fonts.get(typeface);
    }
}

// Uso
public class flyweight {
    public static void main(String[] args) {
        FontFactory factory = new FontFactory();

        Font font1 = factory.getFont("Arial");
        Font font2 = factory.getFont("Arial");
        Font font3 = factory.getFont("Times New Roman");

        System.out.println(font1 == font2); // true
        System.out.println(font1 == font3); // false
    }
}

File: 11.flyweight/flyweight.py
# Flyweight
class Font:
    def __init__(self, typeface):
        self.typeface = typeface

# Flyweight Factory
class FontFactory:
    def __init__(self):
        self.fonts = {}

    def get_font(self, typeface):
        if typeface not in self.fonts:
            self.fonts[typeface] = Font(typeface)
        return self.fonts[typeface]

# Uso
factory = FontFactory()

font1 = factory.get_font("Arial")
font2 = factory.get_font("Arial")
font3 = factory.get_font("Times New Roman")

print(font1 is font2)  # True
print(font1 is font3)  # False